### ISSUE TITLE
initial value and big endian start bit problem.

### ISSUE THREAD

**Issue by PUNITPNT** [eerimoq/cantools#333]
hi,
My Requirement:
I am converting DBC to Excel, according to candbc++ editor excel format using python 3.9.
cantools latest version:

Problem 2:
big endian start bit is different from as it present in vector Candbc++ editor.

Problem 3:
i am getting initial value from cantools as NONE , but initial value present in dbc and visible in  vector Candbc++ editor.

**Comment by juleq** [comment]
Problem 1 is not a problem. CANdb++ in its standard setting will display the start of a signal different from what the representation in the dbc file looks like. You can go into the settings and inspect the options there, but I would not recommend changing.

```

// Little endian
//   DBC     Startbit  : Bit numbering -> LSB0, Reference bit -> LSB
//   CANdb++ Startbit  : Bit numbering -> LSB0, Reference bit -> LSB
 
// Big endian
//   DBC     Startbit  : Bit numbering -> LSB0, Reference bit -> MSB
//   CANdb++ Startbit  : Bit numbering -> LSB0, Reference bit -> LSB
 
// Convert bit numbering (both ways)
// b = b - (b % 8) + 7 - (b % 8)
 
// Convert reference bit LSB to MSB (when bit numbering is MSB0)
// b = b + 1 - length
 
// Convert reference bit MSB to LSB (when bit numbering is MSB0)
// b = b + length - 1

```

For problem 2: Some code and a dbc would probably required to help.

**Comment by PUNITPNT** [comment]
hi 

Thanks for information for problem 1 .
I need  to convert cantools big endian start bit (msb to lsb ) to match vector Candbc++ editor value.

problem 2.

some of initial value of signal , I get as none type,but in candbc editor initial value is present.
when value will be of None type ? 

initial value of signal from cantools not match with Candbc++ editor .

how to match can tools value to Candbc++ editor value ?

how to convert physical to raw, considering all types in can tools ?

![image](https://user-images.githubusercontent.com/26304961/129453100-4b65647b-0a77-4be9-a477-193f845d004f.png)

**Comment by andlaus** [comment]
@juleq probably wanted to say that we would appreciate if you boiled down your DBC file to something which shows the problem  and that can be shared freely (i.e., containing no confidential data). If you do so, I volunteer to try to fix the issue and to also create a unit test so that it will not re-appear in the future. (see #311 for an example of how this is supposed to work.)

**Comment by juleq** [comment]
Yes. Further, since you are already debugging, you could inspect the dbc_specifics object. You should find the attribute GenSigStartValue associated with your signal in the attributes there. If so, the issue would be that the initial value attribute is properly read when parsing the dbc, but not applied to the respective signal property when building up the internal db. Please check if that might be the case here.

**Comment by PUNITPNT** [comment]
@juleq There is not attribute defined for the signal.
![image](https://user-images.githubusercontent.com/26304961/129455559-65bf5122-c6c0-42a9-8904-c305826be8bb.png)

but GenSigStartValue is present in attribute_definitions.
![image](https://user-images.githubusercontent.com/26304961/129455584-cb1ee87c-8711-4149-b704-8f0c5e330222.png)

in attribute definition it contain default values.

@andlaus I will try to create sample dbc and try to reproduce the same issue.

**Comment by PUNITPNT** [comment]
hi

Initial value : None
sample dbc attached. I have reproduced the same issue. 
please check I am getting initial value as None. 
[SampleDBC.zip](https://github.com/eerimoq/cantools/files/6988380/SampleDBC.zip)


start bit :
can tools will give startbit as per below
Big endian = > DBC     Startbit  : Bit numbering -> LSB0, Reference bit -> MSB

for conversion as per CANdb++ Startbit , what exactly formula need to use ?

b = startbit from cantools ?

b = b - (b % 8) + 7 - (b % 8)
b = b + length - 1

this not matching with the value of CANdb++ Startbit

**Comment by juleq** [comment]
Try this function to convert the start bit representation from dbc to CANdb++. It is C# but easy to adapt:

```
private static uint GetStartBitInVectorCANdbppStyle(Signal signal)
{
  uint startbit = (uint)signal.Start;
  if (signal.IsBigEndian)
  {
    startbit += 7 - 2 * (startbit % 8);
    startbit += (uint)signal.Length - 1;
    startbit += 7 - 2 * (startbit % 8);
  }
  return startbit;
}
```
Maybe I get a chance to look into your dbc tomorrow.

**Comment by juleq** [comment]
The dbc does contain an attribute definition for the signal type (SG_) that is of value type FLOAT and has a default value of 0. However, there is no signal in the dbc that has this attribute set to it. So either your example is missing that last step or everything is fine. It might be that Vector tools fall back to this default if there is nothing specificly defined for a signal. But I wonder what their logic would be in this case since the signal you have there is an integer type but the default is FLOAT...

![grafik](https://user-images.githubusercontent.com/1632707/129496754-06dfb523-b060-4354-9d8a-8ed5b2c35483.png)

**Comment by PUNITPNT** [comment]
> Try this function to convert the start bit representation from dbc to CANdb++. It is C# but easy to adapt:
> 
> ```
> private static uint GetStartBitInVectorCANdbppStyle(Signal signal)
> {
>   uint startbit = (uint)signal.Start;
>   if (signal.IsBigEndian)
>   {
>     startbit += 7 - 2 * (startbit % 8);
>     startbit += (uint)signal.Length - 1;
>     startbit += 7 - 2 * (startbit % 8);
>   }
>   return startbit;
> }
> ```
> 
> Maybe I get a chance to look into your dbc tomorrow.

Thanks for the information , I understood the formula.

**Comment by PUNITPNT** [comment]
> The dbc does contain an attribute definition for the signal type (SG_) that is of value type FLOAT and has a default value of 0. However, there is no signal in the dbc that has this attribute set to it. So either your example is missing that last step or everything is fine. It might be that Vector tools fall back to this default if there is nothing specificly defined for a signal. But I wonder what their logic would be in this case since the signal you have there is an integer type but the default is FLOAT...
> 
> ![grafik](https://user-images.githubusercontent.com/1632707/129496754-06dfb523-b060-4354-9d8a-8ed5b2c35483.png)

hi ,
Thanks for information.

if GenSigStartValue attribute not set to the signal, 
then Cantools will make initial value as None.

if None then vlaue will be 0: raw =0

so iv =  (raw * factor ) + offset

so iv will become equal to offset 

Is my understanding Correct ?

**Comment by juleq** [comment]
Do you mean when encoding a message without setting the signal value? I could walk through the code to answer that but it is probably easier you try it or even step through the encoding process with your debugger.

**Comment by PUNITPNT** [comment]
@juleq 
yes I will  try with debugger, 

but I want to understand at what condition cantools return initial value as None  ?

**Comment by PUNITPNT** [comment]
I want to write all signal data into excel.

As i am getting signal.initial  value as None. 

![image](https://user-images.githubusercontent.com/26304961/129594374-5bb14175-284e-4638-b909-fc8601edf195.png)





so because of None , nothing is written in excel sheet for that particular signal.

**Comment by juleq** [comment]
I would argue that this is a reasonable behaviour. Vector seems to fall back to the generic attribute which is FLOAT 0, but I think that that is a surprising choice since the signal is of integer type. I don´t really get their reasoning behind that. But it might also be that I am misinterpreting what I saw in CANdb++ and Vector will default to uint 0 when so generic initial integer value is defined.

If you also want to default to 0, just check if value is None and replace that with 0 on export.

**Comment by PUNITPNT** [comment]
@juleq 
ya I understood your explanation and I agree to it.

I just created sample dbc for one signal, as i will try changing the FLOAT and check.

**Comment by PUNITPNT** [comment]
hi @juleq 

I changed the value FLOAT to INT , still i got None .

what I observed is when "GenSigStartValue"  attribute not set to signal, that time cantools return value as None .

and with respect to vector tool , when  "GenSigStartValue"  attribute not set to signal, it fall back to default value defined in attribute definitions.

So whenever I get initial value as none from cantools , then I taking default value of GenSigStartValue from attribute defintion.

**Comment by juleq** [comment]
Sounds reasonable to do so.

But for cantools in general, I would argue that such a generic fallback would not be mandatory. When you do a residual bus simulation you should probably initialize the signal values or define the initial value for each specific signal to avoid surprises due to a generic fallback. But it is a design choice to do one or the other...

**Comment by PUNITPNT** [comment]
Hi,
@juleq with respect to current requirement to display vector data in excel sheet without use of vector tool ,I think this approach it will work.
Thanks for your information.

I have additional requirement to this.
REQ : To get Japanese character as it is from dbc and write Japanese character to excel .

as my dbc will contain some Japanese character in comment attribute.
when I read and write comments attribute  in excel , it will display some unreadable characters for Japanese letter.

Can you tell How to handle this Japanese characters ?

**Comment by juleq** [comment]
I think cantools has a parameter to define encoding if that is an issue. I assume it should be UTF-8 bit check that using a capable editor that you can load your dbc into to find out what encoding it uses.

I would assume that cantools uses utf-8 internally, but have not checked.

I do not know how you output, but I think you could define encoding there to. Your mileage may vary on output, not sure if you use text-based output or some OLE / ODBC stuff to get to excel. 

It might be painful but having a recent Python and its packages at your disposal will certainly help if you need to fix stuff along the way (e.g. I vaguely remember using iconv for some encoding matters...).

**Comment by PUNITPNT** [comment]
hi @juleq
Coding language : python 3.9
editor : Pycharm

I read dbc using cantool package and store in python dictionary.
I use openpyxl to write dictionary data to excel .

Ex:
worksheet.cell(row=i, column=j).value = "Data"

if Cantools use utf 8, then it able to read Japanese character ?

I am getting this from cantools:
![image](https://user-images.githubusercontent.com/26304961/130472761-58c60967-a9bb-48cd-ac77-3bdd47d52539.png)

same thing it is written to excel.

**Comment by juleq** [comment]
Looks like you need to specify the encoding when reading in. Check the docs for cantools to see how to do that.

Find out what encoding to use by opening dbc in editor.

As far as I see utf-8 does contain most Japanese chars, but not all (?). So it might mostly work out of the box as long as you get the encoding parameters right. 

A good resource might be googling for handling of Japanese text with Python in general. You will come across usual encodings and packages for handling such stuff that way.

**Comment by PUNITPNT** [comment]
hi @juleq 

I gone through cantools doc,  I changed encoding to cp932 for Japanese Characters  while reading DBC.

self.db = cantools.database.load_file(self.path,  encoding ="cp932")

Great Cantools took care of encoding part also.

Thanks for your Quick Reply and suggestions.

**Comment by PUNITPNT** [comment]
hi @juleq 

when I testing my dbc to excel script, I observed that,
cantools returns None for Min and Max value ?

at what condition Cantools return None for min and max value ?

EX: For this signal
 SG_ xxxx : 232|5@1+ (1,0) [0|0] "NA"  abc


Note : Vector gives min =0 max =0

**Comment by juleq** [comment]
Cantools is not specific to any db format.

The issue sou are seeing is that dbc represents None with 0 in this case, because its syntax has no other means to represent None.

You can always look into the dump function for dbc to see the way cantools does dbc specific replacements.

**Comment by PUNITPNT** [comment]
hi @juleq 

it means ,
if Cantools return None for min and max value , then its ok to consider 0 
None = 0 

is this ok ?

I observed in vector min = 0  max =0
I got none in cantools,


I observed in vector min = 1  max =1
can tools give min = 1 , max =1

**Comment by juleq** [comment]
Yes, if min and max are 0 that is the dbc files way of saying None (not defined).

**Comment by PUNITPNT** [comment]
hi @juleq 

thanks for clarification . 

I will write 0 to excel if I got min and max as none.
